### ISSUE TITLE
false negative on use `env` in `env` definition

### ISSUE THREAD

**Issue by kazuk** [rhysd/actionlint#158]
this commit passes `actionlint`

https://github.com/kazuk/sbom-ghr/pull/2/commits/fe18f2d3e424548e791ab9335ea4bc63e662a4e4

but `The workflow is not valid. ` shown as bellow.

```
The workflow is not valid. .github/workflows/rust.yml (Line: 18, Col: 18): Unrecognized named-value: 'env'. Located at position 1 within expression: env.KEY_PREFIX .github/workflows/rust.yml (Line: 43, Col: 22): Unrecognized named-value: 'env'. Located at position 1 within expression: env.KEY_PREFIX
```

**Comment by rhysd** [comment]
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#env

> Variables in the env map cannot be defined in terms of other variables in the map.

Official document mentioned this limitation here.

https://docs.github.com/en/actions/learn-github-actions/contexts#env-context

> You can use the env context in the value of any key in a step except for the id and uses keys. 

`id` and `uses` also seem to ban `env` context.

**Comment by rhysd** [comment]
Now workflow like below

```yaml
on: push

jobs:
  my_job:
    runs-on: ubuntu-latest
    env:
      FOO: aaa
    steps:
      - uses: test/${{ env.FOO }}@main
        env:
          BAR: ${{ env.FOO }}
        id: foo-${{ env.FOO }}
```

causes errors as follows:

```
test.yaml:9:24: undefined variable "env". available variables are "github", "inputs", "job", "matrix", "needs", "runner", "secrets", "steps", "strategy" [expression]
  |
9 |       - uses: test/${{ env.FOO }}@main
  |                        ^~~~~~~
test.yaml:11:20: undefined variable "env". available variables are "github", "inputs", "job", "matrix", "needs", "runner", "secrets", "steps", "strategy" [expression]
   |
11 |           BAR: ${{ env.FOO }}
   |                    ^~~~~~~
test.yaml:12:21: undefined variable "env". available variables are "github", "inputs", "job", "matrix", "needs", "runner", "secrets", "steps", "strategy" [expression]
   |
12 |         id: foo-${{ env.FOO }}
   |                     ^~~~~~~
```

**Comment by rhysd** [comment]
This fix was included in the latest release [v1.6.14](https://github.com/rhysd/actionlint/releases/tag/v1.6.14).

**Comment by MatthewRyanRead** [comment]
On v1.6.14 I am seeing a lot of errors around uses around `env.SOMETHING`.  It seems to be cases where the variables may be defined on a custom runner and aren't runner defaults or initialized earlier in the workflow.

**Comment by rhysd** [comment]
@MatthewRyanRead Do you mean referring `env` context in `env:` is allowed when it is run on self-hosted runner? Could you show the link to the actual workflow if it is public?

**Comment by guidojw** [comment]
Since these changes, I am experiencing a problem accessing top-level `env` in a job step's `env`.

Commit: https://github.com/csvalpha/sofia/pull/750/commits/51db18698b5e426d7d14efe586a8189f2e2907ba

The [job configuration](https://github.com/csvalpha/sofia/blob/d4eef4d6bad51befc225e27ef41e7a23a92e92a0/.github/workflows/continuous-delivery.yml#L173) worked before, so this new error seems to be incorrect.

**Comment by wwalser** [comment]
I too was referencing env.GITHUB_SHA in the end: declarations of an individual task. That is now reporting as invalid.

**Comment by guidojw** [comment]
@rhysd Is this on your radar or should I open a new issue for it?

**Comment by rhysd** [comment]
Thank you for the information.

I understood what is wrong.

- `env:` at job level does not allow to refer `env` econtext
- but `env:` at step level allows

I'll fix this and soon release 1.6.15

**Comment by rhysd** [comment]
I added the fix at 323ceb4.

**Comment by rhysd** [comment]
The fix was released with [v1.6.15](https://github.com/rhysd/actionlint/releases/tag/v1.6.15).
